### PR TITLE
[MHLO]: refine delete redundant specialization function

### PR DESCRIPTION
**Pull request by lipracer**
using c++ SFINAE auto select map template paramater

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h [hunk 1]
```diff
@@ -165,15 +246,64 @@ struct isComplexType {
...
 // Inserts the computation that corresponds to the body of the loop for lowered
 // LHLO unary/binary op. Returns the value for the result.
 template <typename LhloOpTy>
 inline Value MapLhloOpToStdScalarOp(Location loc, ArrayRef<Type> result_types,
                                     ArrayRef<Type> arg_types,
                                     ArrayRef<Value> args, OpBuilder* b) {
-  return MapLhloOpToScalarOpImpl<isSignedIntegerType, ScalarIOp<LhloOpTy>,
-                                 isUnsignedIntegerType, ScalarUOp<LhloOpTy>,
-                                 isFloatType, ScalarFOp<LhloOpTy>>{}(
+  // TODO(lipracer): we need check AllTemplateArguments at compile time
```
Comment (River707) [inline tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h:299]: Have you considered something simpler than this? The template meta-programming here is a bit hard to follow at first glance. For example, what if we did something like(might not be 100% exact):

```c++
struct LhloToScalarOpBase {
  using FOp = void;
  using IOp = void;
  using UOp = void;
  using COp = void; 
};

template <typename LhloBinaryOpTy>
struct LhloToScalarOp : public LhloToScalarOpBase {};
template <>
struct LhloToScalarOp<lmhlo::AddOp> : public LhloToScalarOpBase {
  using FOp = ::mlir::AddFOp;
  using IOp = ::mlir::AddIOp;
  using UOp = ::mlir::AddIOp;
  using COp = ::mlir::complex::AddOp;
};
template <>
struct LhloToScalarOp<lmhlo::AndOp> {
  using IOp = ::mlir::AndOp;
  using UOp = ::mlir::AndOp;
};
...

// Note the `void` middle param, which handles the case where the Op didn't
// provide the corresponding FOp/IOp/etc.
template <typename SupportedType, typename... Args>
struct MapLhloOpToScalarOpImpl<SupportedType, void, Args...> {
  Value operator()(Location loc, ArrayRef<Type> result_types,
                   ArrayRef<Type> arg_types, ArrayRef<Value> args,
                   OpBuilder* b) {
    return MapLhloOpToScalarOpImpl<Args...>{}(loc, result_types, arg_types,
                                              args, b);
  }
};

// Inserts the computation that corresponds to the body of the loop for lowered
// LHLO unary/binary op. Returns the value for the result.
template <typename LhloOpTy>
inline Value MapLhloOpToStdScalarOp(Location loc, ArrayRef<Type> result_types,
                                    ArrayRef<Type> arg_types,
                                    ArrayRef<Value> args, OpBuilder* b) {
  return MapLhloOpToScalarOpImpl<
    isSignedIntegerType, ScalarIOp<LhloOpTy>, 
    isUnsignedIntegerType, ScalarUOp<LhloOpTy>,
    isFloatType, ScalarFOp<LhloOpTy>,
    isComplexType, ScalarCOp<LhloOpTy>>(
      loc, result_types, arg_types, args, b);
}
```

Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h]: Done.

**Review by River707 — APPROVED** [review]
Looks much cleaner, nice!

**Comment by lipracer** [comment]
Should I update my code base？

**Comment by joker-eph** [comment]
TensorFlow is for now still limited to C++14 unfortunately, and our linter flagged there:
```
  /third_party/tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h:249 ClangTidy: template template parameter using 'typename' is incompatible with C++ standards before C++17
  /third_party/tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h:254 ClangTidy: template template parameter using 'typename' is incompatible with C++ standards before C++17
```

**Comment by lipracer** [comment]
Thank you for reminding me. I'll fix it later.

**Review by joker-eph — APPROVED** [review]

**Comment by jessecantu** [comment]
> using c++ SFINAE auto select map template paramater

**Comment by lipracer** [comment]
> > using c++ SFINAE auto select map template paramater

My original intention is to simplify [here](https://github.com/tensorflow/tensorflow/blob/087d5ba662df9265f47bb958ef6ffa767b8707a0/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc#L2804). We can use Mapable to generate a sequence and filter out Ops that cannot be Map. Unfortunately, other Passes have references [like this](https://github.com/tensorflow/tensorflow/blob/087d5ba662df9265f47bb958ef6ffa767b8707a0/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/lhlo_legalize_roots_to_loops.cc#L69).It is assumed that all Ops can be Mapable. In fact, some Ops will get null and there is a risk of Crash.
